#16746 Change required class

vajo3694 opened the pull request.

Commit:
#16746 Change required class

dereuromark commented on src/View/Helper/FormHelper.php:
```diff
@@ -117,9 +117,9 @@ class FormHelper extends Helper
             'inputSubmit' => '<input type="{{type}}"{{attrs}}/>',
             // Container element used by control().
-            'inputContainer' => '<div class="input {{type}}{{required}}">{{content}}</div>',
+            'inputContainer' => '<div class="input {{type}} {{required}}">{{content}}</div>',
```
I dont think you should change the whitespace here, it has other side effects beyond your expected change.

markstory replied:
Agree, this template should not change.

ghost commented:
Ok, maybe silly question but may I change tests for this or should I try another way? Thanks for answer.

ndm2 reviewed (commented):
I'd say that this should be done in a backwards compatible way, even if it would go into `4.next`. Changing HTML output is always a pain for applications as well as plugins, it will affect all sorts of things, tests, CSS, JavaScript, etc.

IMHO the default should stay `required`, and empty as the inverse. Configuring the class could easily be made possible by adding an option similar to `errorClass`.

Representing inverse states feels kinda weird, I don't think I've ever seen something like that in CSS for states that are binary. However if people think that is something that should be supported, then I'd suggest that the question should be asked whether this should be supported for other states too, for example error states, as having it apply to only the required state seems rather arbitrary.

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1180,7 +1180,7 @@ protected function _inputContainerTemplate(array $options): string
             'content' => $options['content'],
             'error' => $options['error'],
-            'required' => $options['options']['required'] ? ' required' : '',
+            'required' => $options['options']['required'] ? 'is-required' : 'not-required',
```
We can't change these default values either, as it will break existing applications. In the issue you opened it sounded more like you were hoping to add a template *for* the required class much like how `selectedClass` is implemented.

markstory commented:
> I'd say that this should be done in a backwards compatible way, even if it would go into 4.next. Changing HTML output is always a pain for applications as well as plugins, it will affect all sorts of things, tests, CSS, JavaScript, etc.

Agreed. I won't merge a change that changes the default HTML output for 4.x. I don't even think we have a solid reason to do that in 5.x. CSS class name conventions involve a ton of _preference_, and the defaults we have are the ones that all existing applications are currently using.

> IMHO the default should stay required, and empty as the inverse. Configuring the class could easily be made possible by adding an option similar to errorClass.

Yes, using a solution similar to `selectedClass` would preserve backwards compatibility and enable users to change the `requiredClass` to fit their preferences whether that value is ` --required` or `must-haz` application authors can do what they need without breaking compatibility.

ghost commented:
I understand and agree with solution on add class option instead change the current output. You’re definitely right about backward compatibility and I apologise for my hasty pull request. Thanks for yours patience.

> Message ID: ***@***.***>

Commit:
Merge branch '4.x' into 4.x-form-input-container-template-required-class

Commit:
#16746 Change required class

Commit:
#16746 Change required class

ndm2 commented on src/View/Helper/FormHelper.php:
```diff
@@ -150,6 +150,8 @@ class FormHelper extends Helper
             // selected class
             'selectedClass' => 'selected',
+            // required class
+            'requiredClass' => 'required',
```
Why does this need to be a template?

I guess I have to generally ask why classes exist as templates. I've never really noticed the `selectedClass` template before 🤔

markstory replied:
The use-case is users using CSS methodologies like BEM and wanting to make the HTML generated by CakePHP to conform to those conventions.

ndm2 replied:
Sure, I get why one would want to be able to change the class, I'm just wondering why this would need to be a template instead of an option like `errorClass`?

I'm already confused by `selectedClass` being a template, it looks workaround-ish to make it available to widgets, but that could've also been achieved by passing an option into the widgets `render()` method?

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1176,11 +1178,15 @@ protected function _inputContainerTemplate(array $options): string
         if (!$this->templater()->get($inputContainerTemplate)) {
             $inputContainerTemplate = 'inputContainer' . $options['errorSuffix'];
         }
+        $requiredClassTemplate = $this->_defaultConfig['templates']['requiredClass'];
+        if ($this->templater()->get('requiredClass')) {
+            $requiredClassTemplate = $this->templater()->get('requiredClass');
+        }
```
```suggestion
        $requiredClassTemplate = $this->templater()->get('requiredClass');
```

I don't think the condition adds anything, and if users want to remove the required class from their templates, they are free to do that. :shrug:

jarovarga replied:
It's for default config class `required`. For change just set template in AppView (tested and working):

```
$this->Form->setTemplates([
    'inputContainer' => /** @lang text */ '<div class="form__input --{{type}}{{required}}">{{content}}</div>',
    'requiredClass' => '--required',
]);
```

jarovarga replied:
Because class for backward comp. must be `required` as default. And if I want change default for my need then set custom class in my AppView.

markstory replied:
Sure, but the default template set also includes a template for `requiredClass` which would allow `get('requiredClass')` to work all of the time.

jarovarga replied:
`$this->templater()->get('requiredClass');` without set template var in AppView return nothing :( I don't understant

markstory replied:
Aren't you including the `requiredClass` in the `templates` configuration data? That is how all the other templates are defaulted and they are all read with `templater->get()`. This behaving differently means something is wrong with the current implementation.

othercorey replied:
Any more information on this?

jarovarga replied:
It's done as @markstory sugested...

Arhell committed:
update src folder links

markstory committed:
Merge pull request #16770 from Arhell/upd-links

update src folder links

Arhell committed:
update unicode links

ADmad committed:
Merge pull request #16772 from Arhell/fix-unicode

update unicode links

Commit:
Merge branch '4.x' into 4.x-form-input-container-template-required-class

Commit:
#16746 Change required class

dependabot[bot] committed:
Bump actions/stale from 5 to 6

Bumps [actions/stale](https://github.com/actions/stale) from 5 to 6.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

ADmad committed:
Merge pull request #16774 from cakephp/dependabot/github_actions/actions/stale-6

Bump actions/stale from 5 to 6

markstory reviewed (commented):
Looking good, we'll need a test covering that changing the `requiredClass` also modifies the generated HTML and then we should be good to merge this 😄

ADmad commented:
Shouldn't this target 4.next?

othercorey committed:
Reset permissions block for api-docs

othercorey committed:
Add permissions block back to api-docs

markstory committed:
Add test for #16779

I was not able to reproduce the problem locally.

markstory committed:
Update version number to 4.4.6

markstory committed:
Add array cast to result of debugInfo to fix iteration issues.

markstory committed:
Merge pull request #16781 from cakephp/issue-16779

Add test for #16779

dependabot[bot] committed:
Bump styfle/cancel-workflow-action from 0.10.0 to 0.10.1

Bumps [styfle/cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) from 0.10.0 to 0.10.1.
- [Release notes](https://github.com/styfle/cancel-workflow-action/releases)
- [Commits](https://github.com/styfle/cancel-workflow-action/compare/0.10.0...0.10.1)

---
updated-dependencies:
- dependency-name: styfle/cancel-workflow-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

othercorey committed:
Update .github/workflows/cancel.yml

Co-authored-by: ADmad <ADmad@users.noreply.github.com>

othercorey committed:
Merge pull request #16784 from cakephp/dependabot/github_actions/styfle/cancel-workflow-action-0.10.1

Bump styfle/cancel-workflow-action from 0.10.0 to 0.10.1

ADmad committed:
Ignore error reported by phpstan

ADmad committed:
Rename phpstan.neon to phpstan.neon.dist

This allows devs to have their own phpstan.neon with user specific options
like "editorUrl", https://phpstan.org/user-guide/output-format#opening-file-in-an-editor

ADmad committed:
Merge pull request #16793 from cakephp/phpstan

Ignore error reported by phpstan

ADmad committed:
Update psalm

othercorey committed:
Merge pull request #16794 from cakephp/phpstan-config-file

Rename phpstan.neon to phpstan.neon.dist

othercorey committed:
Merge pull request #16795 from cakephp/psalm

Update psalm

markstory committed:
Add test for propertyName and validation

I was not able to reproduce the issue in #16778

ndm2 committed:
4.x - Fix cancel action constraint.

Partial constraints aren't supported.

dereuromark committed:
Merge pull request #16803 from cakephp/ndm2-patch-1

4.x - Fix cancel action constraint.

Tyl13 committed:
Correcting documentation for Xml::build

The Xml::build from local file requires readFile to be true, but is not stated in the PHP Docblock example.

markstory committed:
Merge pull request #16804 from Tyl13/patch-1

Correcting documentation for Xml::build

markstory committed:
Merge pull request #16797 from cakephp/issue-16778

Add test for propertyName and validation

markstory commented:
@jarovarga did you want a hand with the tests for this change?

ghost commented:
Hi, yes please. I wrote you via Discord DM.

> Message ID: ***@***.***>

Commit:
Merge branch '4.x' into 4.x-form-input-container-template-required-class

othercorey commented:
@markstory Going to assign this to you to close out

markstory commented:
I picked out the relevant changes into 87a8e1f09926e24fe9736119273f969cb24449df and added tests in 17c40ee2275341058d22ac383338a5375e36d463
